base/iterators.jl

```diff
@@ -1547,7 +1547,9 @@ Stacktrace:
 [...]
 ```
 """
-@propagate_inbounds function only(x)
+@propagate_inbounds only(x) = only(x, Val(:iterate))
+
+@propagate_inbounds function only(x, ::Val{:iterate})
```

Comment:
I'm asking myself whether these versions should be (semi)public (under the current or a different name). Currently the method is chosen on a case-by-case basis. Package developers and other users may want to do the same.

This also reflects that currently there is no general principle to choose the method. It's tempting to use `haslength`, but that doesn't always do the right thing:
```
typeof(y) = String
Base.haslength(typeof(y)) = true
  2.652 μs (0 allocations: 0 bytes)   # runtest with only(y, Val(:iterate))
  3.185 μs (0 allocations: 0 bytes)   # runtest with only(y, Val(:first))
typeof(y) = Vector{Char}
Base.haslength(typeof(y)) = true
  913.744 ns (0 allocations: 0 bytes)
  961.190 ns (0 allocations: 0 bytes)
typeof(y) = BitSet
Base.haslength(typeof(y)) = true
  4.046 μs (0 allocations: 0 bytes)
  5.834 μs (0 allocations: 0 bytes)
```
Maybe a function `hasfastlength` in analogy with `hasfastin` would help. On the other hand, `only` is probably not important enough to justify major additions.

Any thoughts?

Reply:
The new method should also be used for `WeakKeyDict` and maybe also for `PersistentDict` and `Pairs`. For `ImmutableDict` there is a mysterious allocation that slows it down:
```
typeof(y) = WeakKeyDict{Vector{Any}, Int64}
Base.haslength(typeof(y)) = true
  425.247 μs (9000 allocations: 203.12 KiB)   # runtest with only(y, Val(:iterate))
  253.297 μs (6000 allocations: 156.25 KiB)   # runtest with only(y, Val(:first))
typeof(y) = Base.PersistentDict{Char, Int64}
Base.haslength(typeof(y)) = true
  1.455 ms (18000 allocations: 562.50 KiB)
  1.203 ms (16000 allocations: 500.00 KiB)
typeof(y) = Base.ImmutableDict{Char, Int64}
Base.haslength(typeof(y)) = true
  1.594 μs (0 allocations: 0 bytes)
  2.715 μs (1 allocation: 32 bytes)
typeof(y) = Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{x::Int64}}
Base.haslength(typeof(y)) = true
  9.510 μs (0 allocations: 0 bytes)
  9.261 μs (0 allocations: 0 bytes)
```

Reply:
As new public API, there are a few issues to work out here and this would require significantly more technical review since we would be committing to this specific interface for quite some time.

Additionally, rather than using `Val` the typical pattern here for the API would be to create trait types such as `Base.HasLength`.

Maybe something like as follows:

```julia
abstract type OnlyApproach end
struct OnlyIterate <: OnlyApproach end
struct OnlyFirst <: OnlyApproach end
```

Reply:
Rather than using `Val` what if we used the functions themselves as the second argument.

```julia
function only(x, ::typeof(iterate))
   # ...
end

function only(x, ::typeof(first)
    # ...
end
```

One can then invoke the functions as `only(x, iterate)` or `only(x, first)`.